### PR TITLE
Add spack audit packages-https

### PR DESCRIPTION
**Pull request by vsoch**
This PR will add a new audit, specifically for spack package homepage urls (and eventually other kinds I suspect) to see if there is an http address that can be changed to https. Usage is as follows:

```bash
$ spack audit packages-https <package>
```
And in list view:

```bash
$ spack audit list
generic:
  Generic checks relying on global variables

configs:
  Sanity checks on compilers.yaml
  Sanity checks on packages.yaml

packages:
  Sanity checks on specs used in directives

packages-https:
  Sanity checks on https checks of package urls, etc.
```

I think it would be unwise to include with packages, because when run for all, since we do requests it takes a long time. I also like the idea of more well scoped checks - likely there will be other addresses for http/https within a package that we eventually check. For now, there are two error cases - one is when an https url is tried but there is some SSL error (or other error that means we cannot update to https):

```bash
$ spack audit packages-https zoltan
PKG-HTTPS-DIRECTIVES: 1 issue found
1. Error with attempting https for "zoltan": 
    <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: Hostname mismatch, certificate is not valid for 'www.cs.sandia.gov'. (_ssl.c:1125)>
```
This is either not fixable, or could be fixed with a change to the url or (better) contacting the site owners to ask about some certificate or similar.

The second case is when there is an http that needs to be https, which is a huge issue now, but hopefully not after this spack PR.

```bash
$ spack audit packages-https xman
Package "xman" uses http but has a valid https endpoint.
```

And then when a package is fixed:

```bash
$ spack audit packages-https zlib
PKG-HTTPS-DIRECTIVES: 0 issues found.
```
And that's mostly it. :)

### CODE REVIEW

File: lib/spack/spack/audit.py [hunk 1]
```diff
@@ -37,12 +37,19 @@ def _search_duplicate_compilers(error_cls):
...
 
 try:
     from collections.abc import Sequence  # novm
 except ImportError:
     from collections import Sequence
 
+try:
+    from urllib.request import urlopen
+except ImportError:
+    from urllib2 import urlopen  # type: ignore  # novm
```
Comment (alalazo) [inline lib/spack/spack/audit.py]: Minor: can we use `six.moves.urrlib.request` for this?

File: lib/spack/spack/audit.py [hunk 2]
```diff
@@ -261,6 +268,45 @@ def _search_duplicate_specs_in_externals(error_cls):
     kwargs=('pkgs',)
 )
 
+#: Sanity checks on linting
+# This can take some time, so it's run separately from packages
```
Comment (alalazo) [inline lib/spack/spack/audit.py:269]: :100:

**Review by alalazo — COMMENTED** [review]
I just checked core changes, and they LGTM.  I agree that checking https should be separate from package dsl. Pinging @tldahlgren as I think she submitted a similar draft PR recently, so she may want to have a look here. Ping me in a few days for a merge if there are no further comments.

Also, I'm curious what do you think about the internal audits API, since you are the first person, besides me, to add one Let me know if you think it should be improved somewhere :slightly_smiling_face:

**Comment by vsoch** [comment]
> Also, I'm curious what do you think about the intenal audits API, since you are the first person, besides me, to add one Let me know if you think it should be improved somewhere slightly_smiling_face

From a development standpoint it was so easy to understand! I think I figured out how it worked almost immediately, and then it was just about deciding how to go about adding the command (a new audit group? Same arguments as packages? etc.) So I think you did a really good job designing it.  

In terms of name, I think it could also have been called `spack check`, because auditing makes me (for some reason) think of taxes, and something more relaxed to checking for accuracy than a general check :laughing:.  It feels a little similar to `spack style` in that it's checking the formatting of something, but it's probably different because it's flexible to do non-style things too.

But anyway, for the internal APIs, I don't think anything warrants changing. It would be cool to think about what kinds of other things could benefit from auditing, and then to think about how it could be easy for an organization/spack install to write/deploy custom audits of things, sort of along the lines of the extension stuff? E.g., like packages that are modular, but for audit and other things.

**Comment by alalazo** [comment]
> But anyway, for the internal APIs, I don't think anything warrants changing. It would be cool to think about what kinds of other things could benefit from auditing, and then to think about how it could be easy for an organization/spack install to write/deploy custom audits of things, sort of along the lines of the extension stuff?

That would probably fit into a refactor of custom extensions. I can see that people may want to code spack extensions with new audits, or even custom hooks.

**Review by alalazo — DISMISSED** [review]
I didn't check every package, but core changes look good. I'll be posting this in Slack and merge in a couple of days if nobody has requests / comments.

**Comment by tldahlgren** [comment]
Cool.  I wasn't even aware of the `spack audit` command!  I'll have to see how this impacts what I was doing with #25672 , which clearly overlaps on the homepage front.

**Review by tgamblin — APPROVED** [review]

**Review by tgamblin — CHANGES_REQUESTED** [review]
This looks great!

I have 3 thoughts on this:

1. I think it should be `spack audit package-https [--all] [pkg ...]`, so that we can check a package (or several) at a time.
2. What we've typically done for things that are slow by default is to add a `--all` argument to make sure people know what they're getting into. I think it should either take a package argument or `--all` and print help if neither is provided.
3. Needs a test - coverage is only 38%

I think (1) will help with adding this to CI -- we would  ideally check, for every PR:
- that every package has an https homepage (which I guess is something repology wants)
- that every download works over https and checksums to what we expect (basically validate the checksum over https on each PR)

These things are for another PR but (1) sets us up for them.

**Comment by tldahlgren** [comment]
> This looks great!
> 
> I have 3 thoughts on this:
> 
> 1. I think it should be `spack audit package-https [--all] [pkg ...]`, so that we can check a package (or several) at a time.
> 2. What we've typically done for things that are slow by default is to add a `--all` argument to make sure people know what they're getting into. I think it should either take a package argument or `--all` and print help if neither is provided.
> 3. Needs a test - coverage is only 38%
> 
> I think (1) will help with adding this to CI -- we would ideally check, for every PR:
> 
> * that every package has an https homepage (which I guess is something repology wants)
> * that every download works over https and checksums to what we expect (basically validate the checksum over https on each PR)
> 
> These things are for another PR but (1) sets us up for them.

FWIW. I concur with Todd's three points.

**Comment by vsoch** [comment]
`--all` and tests are added.

**Review by tgamblin — DISMISSED** [review]

**Review by alalazo — APPROVED** [review]

**Pull request merged**